src/main/java/org/breedinginsight/brapps/importer/controllers/UploadController.java

```diff
@@ -140,7 +140,60 @@ public HttpResponse<Response<ImportResponse>> previewData(@PathVariable UUID pro
+            Response<ImportResponse> response = new Response(result);
+            return HttpResponse.ok(response).status(HttpStatus.ACCEPTED);
+        } catch (DoesNotExistException e) {
+            log.error(e.getMessage(), e);
+            return HttpResponse.notFound();
+        } catch (AuthorizationException e) {
+            log.error(e.getMessage(), e);
+            return HttpResponse.status(HttpStatus.FORBIDDEN, e.getMessage());
+        } catch (UnprocessableEntityException e) {
+            log.error(e.getMessage(), e);
+            return HttpResponse.status(HttpStatus.UNPROCESSABLE_ENTITY, e.getMessage());
+        } catch (HttpStatusException e) {
+            log.error(e.getMessage(), e);
+            return HttpResponse.status(e.getStatus(), e.getMessage());
+        }
+    }
+
+    @Put("programs/{programId}/import/mappings/{mappingId}/workflows/{workflow}/data/{uploadId}/preview")
+    @Produces(MediaType.APPLICATION_JSON)
+    @AddMetadata
+    @ProgramSecured(roles = {ProgramSecuredRole.BREEDER, ProgramSecuredRole.SYSTEM_ADMIN})
+    public HttpResponse<Response<ImportResponse>> previewData(@PathVariable UUID programId, @PathVariable UUID mappingId,
+                                                              @PathVariable String workflow, @PathVariable UUID uploadId) {
+        try {
+            AuthenticatedUser actingUser = securityService.getUser();
+            ImportResponse result = fileImportService.updateUpload(programId, uploadId, workflow, actingUser, null, false);
+            Response<ImportResponse> response = new Response(result);
+            return HttpResponse.ok(response).status(HttpStatus.ACCEPTED);
+        } catch (DoesNotExistException e) {
+            log.error(e.getMessage(), e);
+            return HttpResponse.notFound();
+        } catch (AuthorizationException e) {
+            log.error(e.getMessage(), e);
+            return HttpResponse.status(HttpStatus.FORBIDDEN, e.getMessage());
+        } catch (UnprocessableEntityException e) {
+            log.error(e.getMessage(), e);
+            return HttpResponse.status(HttpStatus.UNPROCESSABLE_ENTITY, e.getMessage());
+        } catch (HttpStatusException e) {
+            log.error(e.getMessage(), e);
+            return HttpResponse.status(e.getStatus(), e.getMessage());
+        }
+    }
+
+    @Put("programs/{programId}/import/mappings/{mappingId}/workflows/{workflow}/data/{uploadId}/commit")
+    @Produces(MediaType.APPLICATION_JSON)
+    @AddMetadata
+    @ProgramSecured(roles = {ProgramSecuredRole.BREEDER, ProgramSecuredRole.SYSTEM_ADMIN})
+    public HttpResponse<Response<ImportResponse>> commitData(@PathVariable UUID programId, @PathVariable UUID mappingId,
+                                                             @PathVariable String workflow, @PathVariable UUID uploadId,
+                                                             @Body @Nullable Map<String, Object> userInput) {
+        try {
```

Comment:
Changing `workflow` to `workflowId` would be more consistent with the use here and with the enum.

Reply:
Pushed changes